src/widgets/breadcrumb/__tests__/breadcrumb-testing-library-test.js

```diff
@@ -0,0 +1,286 @@
+
+    search.start();
+
+    await wait();
+
+    search.removeWidget(widget);
+
+    await wait();
+
+    expect(container).toBeEmpty();
+  });
+
+  describe('without initial refinements', () => {
+    it('must reflect navigation within a hierarchical menu', async () => {
+      const hierarchicalMenuContainer = document.createElement('div');
+
+      search.addWidget(
+        instantsearch.widgets.breadcrumb({
+          container,
+          attributes: [
+            'hierarchicalCategories.lvl0',
+            'hierarchicalCategories.lvl1',
+            'hierarchicalCategories.lvl2',
+          ],
+        })
+      );
+
+      search.addWidget(
+        instantsearch.widgets.hierarchicalMenu({
+          showParentLevel: false,
+          container: hierarchicalMenuContainer,
+          attributes: [
+            'hierarchicalCategories.lvl0',
+            'hierarchicalCategories.lvl1',
+            'hierarchicalCategories.lvl2',
+          ],
+          rootPath: 'Cameras & Camcorders',
+        })
+      );
+
+      search.start();
+
+      await wait();
+
+      queryByText(hierarchicalMenuContainer, 'Digital Cameras').click();
+
+      await wait();
+
+      expect(queryByText(container, 'Home')).toBeTruthy();
+      expect(queryByText(container, 'Digital Cameras')).toBeTruthy();
+    });
+  });
+
+  describe('with initial refinements', () => {
+    beforeEach(() => {
+      search.addWidget(
+        instantsearch.widgets.configure({
+          hierarchicalFacetsRefinements: {
+            'hierarchicalCategories.lvl0': [
+              'Cameras & Camcorders > Digital Cameras',
+            ],
+          },
+        })
+      );
+    });
+
+    it('must render the categories', async () => {
+      search.addWidget(
+        instantsearch.widgets.breadcrumb({
+          container,
+          attributes: [
+            'hierarchicalCategories.lvl0',
+            'hierarchicalCategories.lvl1',
+            'hierarchicalCategories.lvl2',
+          ],
+        })
+      );
+
+      search.start();
+
+      await wait();
+
+      expect(queryByText(container, 'Home')).toBeTruthy();
+      expect(queryByText(container, 'Cameras & Camcorders')).toBeTruthy();
+      expect(queryByText(container, 'Digital Cameras')).toBeTruthy();
+    });
+
+    it('must go one level up on click on a category', async () => {
+      search.addWidget(
+        instantsearch.widgets.breadcrumb({
+          container,
+          attributes: [
+            'hierarchicalCategories.lvl0',
+            'hierarchicalCategories.lvl1',
+            'hierarchicalCategories.lvl2',
+          ],
+        })
+      );
+
+      search.start();
+
+      await wait();
+
+      queryByText(container, 'Cameras & Camcorders').click();
+
+      await wait();
+
+      expect(queryByText(container, 'Home')).toBeTruthy();
+      expect(queryByText(container, 'Cameras & Camcorders')).toBeTruthy();
+      expect(queryByText(container, 'Digital Cameras')).toBeNull();
```

Comment:
It could be nice to snapshot the calls to `.search` during each interaction.
One of the issues we had with the hierarchicalMenu is that it triggered too many requests

Reply:
Adding an assertion with the number of requests (i.e. how many times the `searchClient.search()` method was called) might make more sense than snapshotting its calls. In the end, this testing strategy is about what is the state of the app at each step, not necessarily how we got there (this can be considered as an implementation detail). Keeping track of the number of requests can still be an important metric because it has a business impact and should be considered as a feature.

Reply:
I think it makes sense to count and/or snapshots API calls if we consider these requests as a side effect from the widgets.

This could simply be asserted during a test with something like:

```js
// ...

search.start();

await wait();

expect(searchClient.search).toHaveBeenCalledTimes(1);
expect(searchClient.search.mock.calls[0]).toMatchSnapshot();

queryByText(container, 'Home').click();

await wait();

expect(searchClient.search).toHaveBeenCalledTimes(2);
expect(searchClient.search.mock.calls[1]).toMatchSnapshot();